apm-sniffer/apm-test-tools/src/test/java/org/apache/skywalking/apm/plugin/ArbitrarySetTest.java

```diff
@@ -0,0 +1,238 @@
+ * # Benchmark mode: Throughput, ops/time
+ * # Benchmark: org.apache.skywalking.apm.plugin.ArbitrarySetTest.treeMap
+ * <p>
+ * # Run progress: 66.67% complete, ETA 00:03:02
+ * # Fork: 1 of 2
+ * # Warmup Iteration   1: 149.772 ops/ms
+ * # Warmup Iteration   2: 164.658 ops/ms
+ * # Warmup Iteration   3: 181.832 ops/ms
+ * # Warmup Iteration   4: 181.535 ops/ms
+ * Iteration   1: 178.571 ops/ms
+ * Iteration   2: 135.320 ops/ms
+ * Iteration   3: 152.644 ops/ms
+ * Iteration   4: 166.896 ops/ms
+ * Iteration   5: 167.542 ops/ms
+ * <p>
+ * # Run progress: 83.33% complete, ETA 00:01:31
+ * # Fork: 2 of 2
+ * # Warmup Iteration   1: 171.791 ops/ms
+ * # Warmup Iteration   2: 178.640 ops/ms
+ * # Warmup Iteration   3: 167.416 ops/ms
+ * # Warmup Iteration   4: 171.610 ops/ms
+ * Iteration   1: 179.966 ops/ms
+ * Iteration   2: 187.318 ops/ms
+ * Iteration   3: 183.934 ops/ms
+ * Iteration   4: 181.388 ops/ms
+ * Iteration   5: 186.286 ops/ms
+ * <p>
+ * <p>
+ * Result "org.apache.skywalking.apm.plugin.ArbitrarySetTest.treeMap":
+ * 171.986 ±(99.9%) 25.408 ops/ms [Average]
+ * (min, avg, max) = (135.320, 171.986, 187.318), stdev = 16.806
+ * CI (99.9%): [146.578, 197.394] (assumes normal distribution)
+ * <p>
+ * <p>
+ * # Run complete. Total time: 00:09:07
+ * <p>
+ * REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
+ * why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
+ * experiments, perform baseline and negative tests that provide experimental control, make sure
+ * the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
+ * Do not assume the numbers tell you what you want them to tell.
+ * <p>
+ * Benchmark                    Mode  Cnt     Score     Error   Units
+ * ArbitrarySetTest.array      thrpt   10  1678.286 ± 190.386  ops/ms
+ * ArbitrarySetTest.arrayList  thrpt   10   647.654 ±  73.776  ops/ms
+ * ArbitrarySetTest.treeMap    thrpt   10   171.986 ±  25.408  ops/ms
+ *
+ * @author kezhenxu94
+ */
+@BenchmarkMode(Mode.Throughput)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@State(Scope.Thread)
+@Fork(2)
+@Warmup(iterations = 4)
+@Measurement(iterations = 5)
+public class ArbitrarySetTest {
+    private static final Object PLACEHOLDER = new Object();
+
+    @Benchmark
+    public void arrayList() {
+        ArrayList<Object> list = new ArrayList<Object>(Collections.nCopies(20, PLACEHOLDER));
+        for (int i = 0; i < 100; i++) {
+            int oldSize = list.size();
+            if (i >= oldSize) {
+                int newSize = Math.max(oldSize * 2, i);
+                ArrayList<Object> newList = new ArrayList<Object>(newSize);
+                newList.addAll(list);
+                newList.addAll(oldSize, Collections.nCopies(newSize - oldSize, PLACEHOLDER));
+                list = newList;
+            }
+            list.set(i, i);
+        }
+    }
+
+    @Benchmark
+    public void array() {
```

Comment:
You should use a LinkedList as compare testing.

Reply:
Added